checks/basic/docker_pkg_github_com_registry.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2021 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package basic
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/digitalocean/clusterlint/checks"
+	"github.com/digitalocean/clusterlint/kube"
+	"github.com/docker/distribution/reference"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func init() {
+	checks.Register(&dockerPkgGithubComRegistryCheck{})
+}
+
+type dockerPkgGithubComRegistryCheck struct{}
+
+// Name returns a unique name for this check.
+func (l *dockerPkgGithubComRegistryCheck) Name() string {
+	return "docker-pkg-github-com-registry"
+}
+
+// Groups returns a list of group names this check should be part of.
+func (l *dockerPkgGithubComRegistryCheck) Groups() []string {
+	return []string{"basic"}
```

Comment:
I wonder if we should put the check in the `doks` group where we know for sure that it's going to be an issue for users eventually (i.e., when they migrate to 1.20). Other providers/clusters may use CRI-O which I don't know whether it's affected as well. 🤔

Reply:
@timoreimann Is there a way to detect which version you're running the linter against? So we could make it a warning for everything below 1.20, and an error for 1.20 and up.

It's up to you where we should put it, if others are also using `containerd` on their kaas they will run into the same issue. If there is a way to detect if it runs against a DOKS cluster or not, we could do something with that and change the message.

Reply:
>Is there a way to detect which version you're running the linter against? So we could make it a warning for everything below 1.20, and an error for 1.20 and up.

Not right now to my knowledge, although including that information shouldn't be too hard since Kubernetes provides a discovery endpoint with the server version. We could retrieve that information on startup and make it available to all checks.

The 1.20 version is just critical for DOKS: I think the official removal for Docker runtime support only happens in a few releases, but DOKS went ahead and made the switch sooner. So strictly speaking, we should only error out if we're on DOKS 1.20, which we neither provide as an input to clusterlint today but that could be determined heuristically based on certain labels we put on Node objects.

To complete the wish list, it'd be even greater if we could make the target Kubernetes version a configurable parameter to clusterlint: that way, we could not only error out on DOKS 1.20 but also on DOKS 1.19 right before a customer intends to upgrade to the version that's impacted. The latter is the very scenario you ran into a believe, so it'd be beneficial to get that ahead of time warning from clusterlint before things start to fall apart.

So that's a lot I'm asking for, but I don't want to block the PR on dependent fundamental changes needed to reap the biggest benefit -- the check is already very useful on its own. My gut feeling is that we could put it in the DOKS group for now (either as a warning or an error) where we know it's going to apply for sure eventually, and file support tickets to track the discussed work. Once those our complete, the check could be updated to make it apply in more specific contexts.

I'd like to wait for some of my colleagues who get to spend more time with clusterlint before making any decisions, but do let me know what you think.

Reply:
> > Is there a way to detect which version you're running the linter against? So we could make it a warning for everything below 1.20, and an error for 1.20 and up.
> 
> Not right now to my knowledge, although including that information shouldn't be too hard since Kubernetes provides a discovery endpoint with the server version. We could retrieve that information on startup and make it available to all checks.
> 

That would be awesome :+1: . Does it also reveal if `docker` or `containerd` is used as the container runtime? Because if it is, we could make the check even more precise.

> The 1.20 version is just critical for DOKS: I think the official removal for Docker runtime support only happens in a few releases, but DOKS went ahead and made the switch sooner. So strictly speaking, we should only error out if we're on DOKS 1.20, which we neither provide as an input to clusterlint today but that could be determined heuristically based on certain labels we put on Node objects already.
> 

Well thank you for pushing forward in that regard :+1: 

> To complete the wish list, it'd be even greater if we could make the target Kubernetes version a configurable parameter to clusterlint: that way, we could not only error out on DOKS 1.20 but also on DOKS 1.19 right before a customer intends to upgrade to the version that's impacted. The latter is the very scenario you ran into a believe, so it'd be beneficial to get that ahead of time warning from clusterlint before things start to fall apart.
> 

Yes exactly! Warning on the upgrade to 1.19, and error on the upgrade to 1.20. Honestly? I would just not make it possible to upgrade to 1.20 when this check triggers.

> So that's a lot I'm asking for, but I don't want to block the PR on dependent fundamental changes needed to reap the biggest benefit -- the check is already very useful on its own. My gut feeling is that we could put it in the DOKS group for now (either as a warning or an error) where we know it's going to apply for sure eventually, and file support tickets to track the discussed work. Once those our complete, the check could be updated to make it apply in more specific contexts.
> 
> I'd like to wait for some of my colleagues who get to spend more time with clusterlint before making any decisions, but do let me know what you think.

I agree those things should be in the cluster linter, but also agree that with the current impact of DOKS moving ahead with the change we should merge thing in. Probably with a few more changes.